les/utils/limiter.go

```diff
@@ -0,0 +1,405 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+// Limiter protects a network request serving mechanism from denial-of-service attacks.
+// It limits the total amount of resources used for serving requests while ensuring that
+// the most valuable connections always have a reasonable chance of being served.
+type Limiter struct {
+	lock sync.Mutex
+	cond *sync.Cond
+	quit bool
+
+	nodes                                             map[enode.ID]*nodeQueue
+	addresses                                         map[string]*addressGroup
+	addressSelect, valueSelect                        *WeightedRandomSelect
+	maxValue                                          float64
+	maxPriorWeight, totalPriorWeight, totalPriorLimit uint64
+	lastValueSelected                                 bool
```

Comment:
Perhaps `selectedByValue` is a better name?

Reply:
`priorWeight` is confusing for me. `PriorWeight is a rough estimate of the serving cost of the request.`, perhaps we can pick a better name?

And also for `totalPriorWeight`, perhaps we can use `cumulativeXXX`. Because the requests are accumulated

Reply:
I dropped the prior weight vs. variable cost logic in order to make things simpler.